### ISSUE TITLE
[Question] Changing the vision tower.

### ISSUE THREAD

**Issue by ronaldAlvxs** [haotian-liu/LLaVA#1913]
### Question

If I replaced the vision tower with a smaller one — for instance, a CLIP model that outputs 577 features replaced by another that outputs 200 — would it be enough to retrain only the projector, or would the LLM also need to be fine-tuned?
Would the fact that the LLM was originally trained with 577-dimensional visual embeddings affect its inference capability (e.g., through mismatched attention projections)?

**Comment by rehan243** [comment]
Wait, quick question — are you freezing the LLM entirely or just skipping fine-tuning on it? If the new vision tower’s embeddings drop down to 200 dims, the LLM's attention layers won’t know what to do with that mismatch. You’d probably get garbage outputs because it’s hardcoded to expect 577 dims in its input projection weights.

You could technically just retrain the projector to upsample from 200 back to 577 (e.g., a simple linear layer), but in practice, we’ve seen this create bottlenecks if the new embedding space doesn’t align semantically with the old one. If you’re sticking with smaller vision embeddings, I’d recommend fine-tuning the LLM too — even a partial LoRA fine-tune on the input layers should help it adapt.

```python
# Simple projector adjustment example
import torch.nn as nn

projector = nn.Linear(200, 577)  # Map new vision tower output to expected size
projected_features = projector(vision_tower_output)
```

We hit this with a ViT swapped for a custom ResNet once — skipping the LLM fine-tune led to unstable loss during training.
